PC/_wmimodule.cpp

```diff
@@ -0,0 +1,310 @@
+        return NULL;
+    }
+
+    data.query = PyUnicode_AsWideCharString(query, NULL);
+    if (!data.query) {
+        return NULL;
+    }
+
+    if (0 != _wcsnicmp(data.query, L"select ", 7)) {
+        PyMem_Free((void *)data.query);
+        PyErr_SetString(PyExc_ValueError, "only SELECT queries are supported");
+        return NULL;
+    }
+
+    Py_BEGIN_ALLOW_THREADS
+
+    if (!CreatePipe(&data.readPipe, &data.writePipe, NULL, 0)) {
+        err = GetLastError();
+    } else {
+        hThread = CreateThread(NULL, 0, _query_thread, (LPVOID*)&data, 0, NULL);
+        if (!hThread) {
+            err = GetLastError();
+            // Normally the thread proc closes this handle, but since we never started
+            // we need to close it here.
+            CloseHandle(data.writePipe);
+        }
+    }
+
+    while (!err) {
+        if (ReadFile(
+            data.readPipe,
+            (LPVOID)&buffer[offset / sizeof(buffer[0])],
+            sizeof(buffer) - offset,
+            &bytesRead,
+            NULL
+        )) {
+            offset += bytesRead;
+            if (offset >= sizeof(buffer)) {
+                err = ERROR_MORE_DATA;
+            }
+        } else {
+            err = GetLastError();
+        }
+    }
+
+    if (data.readPipe) {
+        CloseHandle(data.readPipe);
+    }
+
+    // Allow the thread some time to clean up
+    switch (WaitForSingleObject(hThread, 1000)) {
+    case WAIT_OBJECT_0:
+        // Thread ended cleanly
+        if (!GetExitCodeThread(hThread, (LPDWORD)&err)) {
+            err = GetLastError();
+        }
+        break;
+    case WAIT_TIMEOUT:
+        // Probably stuck - there's not much we can do, unfortunately
+        if (err == 0 || err == ERROR_BROKEN_PIPE) {
+            err = WAIT_TIMEOUT;
+        }
+        break;
+    default:
+        if (err == 0 || err == ERROR_BROKEN_PIPE) {
+            err = GetLastError();
+        }
+        break;
+    }
+
+    CloseHandle(hThread);
+    hThread = NULL;
+
+    Py_END_ALLOW_THREADS
+
+    PyMem_Free((void *)data.query);
+
+    if (err == ERROR_MORE_DATA) {
+        PyErr_SetString(PyExc_OSError, "Query returns too much data");
+        return NULL;
+    } else if (err) {
+        PyErr_SetFromWindowsErr(err);
+        return NULL;
+    }
+
+    return PyUnicode_FromWideChar(buffer, offset  / sizeof(buffer[0]) - 1);
+}
+
+
+static PyMethodDef wmi_functions[] = {
+    _WMI_EXEC_QUERY_METHODDEF
+    { NULL, NULL, 0, NULL }
+};
+
+static int exec_wmi(PyObject *module)
+{
+    PyModule_AddFunctions(module, wmi_functions);
+
+    return 0; // success
+}
+
+static PyModuleDef_Slot wmi_slots[] = {
+    { Py_mod_exec, exec_wmi },
+    { 0, NULL }
+};
+
+static PyModuleDef wmi_def = {
+    PyModuleDef_HEAD_INIT,
+    "_wmi",
+    NULL, // doc
+    0,              // m_size
+    NULL,           // m_methods
+    wmi_slots,
+    NULL,           // m_traverse
+    NULL,           // m_clear
+    NULL,           // m_free
```

Comment:
For new extensions, I suggest using C99 style:

```suggestion
    .m_name = "_wmi",
    .m_size = 0,
    .m_slots = wmi_slots,
```

Reply:
Unfortunately, this doesn't work because you can't mix initializer syntax with the INIT macro.